### PR TITLE
CLI Batch Processing Tool

### PR DESCRIPTION
**Pull request by dnath**
Command Line Interface -- StochSS Job Batch Processing Tool

### CODE REVIEW

**Comment by bbbales2** [comment]
I'm having trouble getting my credentials to go through.

I tried 3 times, once with an empty db. I enter my credentials to launch a node with stochss (so I can then use that node for these tests) and it's failing.

I'm seeing this error in the logs:
INFO     2015-02-08 20:25:07,363 backendservice.py:429] validateCredentials: exiting with result : <backend.infrastr\
ucture_manager.InfrastructureManager instance at 0x7f393740ff38>
ERROR    2015-02-08 20:25:07,363 backendservice.py:432] validateCredentials: exiting with error : EC2Agent instance \
has no attribute 'validate_credentials

I tested my credentials on another branch and they are valid. Any clues?

**Comment by dnath** [comment]
@Ben, fixed the said issue

**Comment by bbbales2** [comment]
Hey I got an instance running and ran a job. But now I'm not able to check
the output:

{'Message': 'Task sent to Cloud',
 'message': 'Task finished. Generating output.',
 'output': '
https://s3.amazonaws.com/stochss-cli-bucket/output/d170fc22-3c75-4667-8132-129718538e3f.tar
',
 'pid': 'd170fc22-3c75-4667-8132-129718538e3f',
 'start_time': '2015-02-08 14:48:41',
 'status': 'finished',
 'taskid': 'd170fc22-3c75-4667-8132-129718538e3f',
 'time_taken': '0 seconds and 67608 microseconds ',
 'uuid': 'd170fc22-3c75-4667-8132-129718538e3f'}

I try to access that url but get a permission denied. Is there some special
way I should access it with my Amazon keys?

File: run.mac.sh [hunk 1]
```diff
@@ -1,5 +1,12 @@
 #!/usr/bin/env bash
 
+osname=$(uname)
```
Comment (briandrawert) [inline run.mac.sh]: What is the purpose of the changes to these two files?

Comment (bbbales2) [inline run.mac.sh]: It seems like if I run a job via the CLI and then try to run another job
inside StochSS, the job gets stuck in Pending in stochss.

If I just run something on a fresh AMI launched in StochSS it's fine.

Ben

Comment (dnath) [inline run.mac.sh]: @briandrawert I was running ./run.ubuntu.sh on my Mac by mistake, So those lines take care of giving an error when I run the linux StochSS starter script on my Mac and vice versa

@bbbales2 , I forgot to mention because of how celery is configured, you can either run CLI or StochSS from browser and not both at the same time.

Comment (dnath) [inline run.mac.sh]: @bbbales2, Try checking your aws s3 bucket at https://console.aws.amazon.com/s3/home?region=us-east-1 --- Can you access it? If not, there is some issue with granting S3 access to your creds

Comment (bbbales2) [inline run.mac.sh]: I only have command line access.

Is there a way to get these files on the command line?

Ben

Comment (dnath) [inline run.mac.sh]: yes, using the aws command line tools.

One more thing, you can run cli and browser mode at the same time, if you have made two git clones, one running cli and the other browser mode (I have not checked this out -- but celery instances for each would be separate)

Comment (briandrawert) [inline run.mac.sh]: Does this mean that you can not run jobs from the app and jobs from the cli on the same instances?

Comment (dnath) [inline run.mac.sh]: Yes, you can, but can do either run cli over them or use the StochSS ui over them but not both at the same time

**Comment by dnath** [comment]
Done. Check it now.

1. UI and CLI have different celery config files now
2. You can launch and terminate instances through cli

File: app/backend/bin/cli.py [hunk 2]
```diff
@@ -0,0 +1,556 @@
+#!/usr/bin/env python
+
+import sys
+import os
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
+sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'lib', 'boto'))
```
Comment (briandrawert) [inline app/backend/bin/cli.py]: Can you add:
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'lib', 'celery'))

I am getting:
Traceback (most recent call last):
  File "./cli.py", line 26, in <module>
    import tasks
  File "./../tasks.py", line 14, in <module>
    from celery import Celery, group
ImportError: No module named celery

Comment (briandrawert) [inline app/backend/bin/cli.py]: and "kombu", "amqp", "billiard"

File: app/backend/bin/cli.py [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+from tasks import *
+
+from common.config import CeleryConfig, JobTypes, JobDatabaseConfig, JobConfig, AgentTypes, AWSConfig
+import common.helper as helper
+
+import celery
+from celery.task.control import inspect
+
+
+def get_aws_credentials():
```
Comment (briandrawert) [inline app/backend/bin/cli.py]: The environment variables I have are named: AWS_ACCESS_KEY and AWS_SECRET_KEY

It might be more compatible to check both.

**Comment by dnath** [comment]
Check it now.

**Comment by briandrawert** [comment]
There is a bug with the output.  The S3 bucket "stochss-cli-bucket/" does not exist.

$ wget https://s3.amazonaws.com/stochss-cli-bucket/output/4318eed7-45dc-41b7-aafa-ecaa7b98c869.tar
--21:43:34--  https://s3.amazonaws.com/stochss-cli-bucket/output/4318eed7-45dc-41b7-aafa-ecaa7b98c869.tar
           => `4318eed7-45dc-41b7-aafa-ecaa7b98c869.tar'
Resolving s3.amazonaws.com... 54.231.8.144
Connecting to s3.amazonaws.com[54.231.8.144]:443... connected.
HTTP request sent, awaiting response... 403 Forbidden
21:43:34 ERROR 403: Forbidden.

**Comment by bbbales2** [comment]
Yeah that is the same bug I am getting. I get them on cmd line too when I
try to use that folder.

Can you make that folder in the aws ui or what errors do you get when you
try that?

Ben

**Comment by briandrawert** [comment]
The error I get is this:

The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again.

Bucket names are global, we need to add a UUID to the name.  Ideally, this would be the same one as the 'stochss-output-XXXXXX' bucket.

**Comment by dnath** [comment]
Try changing the bucket name in the jobs config --

eg
"output_store": {
        "type" : "amazon_s3",
        "bucket_name" : "stochss-cli-bucketXXXXX"
    },

and see if you get the same error and I'll try to make the bucket with an unique id and make it public --
You can't download it with wget as it does not have aws creds --- but I'm working on making public

**Comment by dnath** [comment]
The buckets now have unique ids + they are public. Check it now.

**Comment by briandrawert** [comment]
Tested and everything is working

**Comment by briandrawert** [comment]
Enhancements:
- flag to auto-download files
- flag to 'auto-configure, run, terminate"
- flag to 'publish' booted instances to the config file (terminate removes machines from config)
- separate machine config and job config files (allow multiple job files)

**Comment by briandrawert** [comment]
Also Enhancements:
- template config file
